### PR TITLE
[dynamo] [dynamic shapes] Fix register_buffer (and all module associated tensors) 

### PR DESCRIPTION
**Pull request by voznesenskym**
We were not associating tensors with symbolic shape guards or creating guards correctly. Introduce logic for adding true TensorRef guards for cases where a tensor is (1) known to the output_graph (2) has symbolic_shapes (3) lives on an nn_module

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/guards.py [hunk 1]
```diff
@@ -159,10 +159,11 @@ def set_export_info(self, guard_type, guarded_class, code_list, obj_weakref):
...
 
-        if not self.code_list:
-            self.code_list = code_list
-        else:
-            self.code_list.extend(code_list)
+        if code_list:
+            if not self.code_list:
+                self.code_list = code_list
+            else:
+                self.code_list.extend(code_list)
```
Comment (ezyang) [inline torch/_dynamo/guards.py:166]: cursed conditional lol

File: torch/_dynamo/guards.py [hunk 2]
```diff
@@ -613,8 +614,18 @@ def combine_scopes(left, right):
         )
         global_builder = GuardBuilder(self.id_ref, f_globals, self, renames=False)
         for guard in sorted(guards or [], key=Guard.sort_key):
-            if not config.guard_nn_modules and guard.is_nn_module():
-                continue
+            if (
+                guard.create_fn == GuardBuilder.TENSOR_MATCH
+                and id(eval(guard.name, local_builder.scope))
```
Comment (ezyang) [inline torch/_dynamo/guards.py:619]: cursed eval

File: torch/_dynamo/guards.py [hunk 3]
```diff
@@ -613,8 +614,18 @@ def combine_scopes(left, right):
...
+                in output_graph.tensor_id_to_sym_shape_ref
+            ):
+                # An exception to not guarding on nn_module properties
+                # In dynamic shapes mode, we sometimes get "weights" "bias" or other registered/named buffers
+                # accesed. We need to install their TENSOR_MATCH guards
+                # so that the names are known for symbolic shape guarding.
+                # This is also probably good for correctness anyway.
+                guard.create(local_builder, global_builder)
+            continue
             guard.create(local_builder, global_builder)
```
Comment (ezyang) [inline torch/_dynamo/guards.py:629]: cursed dead code

File: torch/_dynamo/guards.py [hunk 4]
```diff
@@ -448,26 +449,26 @@ def SYMBOL_MATCH():
     def TENSOR_MATCH(self, guard: Guard):
         if guard.is_nn_module():
             self.ID_MATCH(guard)
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_dynamo/guards.py:451]: this also is probably doing undoing a perf optimization, cc @jansel

[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/guards.py:451]: It only applies the tensor check if it also passes the `guard.is_nn_module` check in guards.py where we `if not config.guard_nn_modules and guard.is_nn_module():`

File: torch/_dynamo/guards.py (hunk 3 above)
Comment (voznesenskym) [reply torch/_dynamo/guards.py:630]: Not dead, fat fingered a line

File: torch/_dynamo/guards.py [hunk 5]
```diff
@@ -614,6 +615,17 @@ def combine_scopes(left, right):
         global_builder = GuardBuilder(self.id_ref, f_globals, self, renames=False)
         for guard in sorted(guards or [], key=Guard.sort_key):
             if not config.guard_nn_modules and guard.is_nn_module():
+                if (
+                    guard.create_fn == GuardBuilder.TENSOR_MATCH
+                    and id(eval(guard.name, local_builder.scope))
```
Comment (voznesenskym) [inline torch/_dynamo/guards.py:620]: This eval might be cursed, but its not new, see: `def get` in this file.

File: torch/_dynamo/guards.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/guards.py:166]: It's alright, tensor_ref just doesn't get a code list.

**Pull request merged**